### PR TITLE
Release notes CI step

### PR DESCRIPTION
**Pull request by vzarytovskii**
Ok, let's make everyone's life miserable and require to fill in release notes every time Compiler, FSharp.Core, LanguageFeatures or VS are changed.

Will help us to generate proper release notes though.

### CODE REVIEW

**Comment by nojaf** [comment]
Can you please add some guidance on what we are expected to write?
Are we following [KeepAChangelog](https://keepachangelog.com/en/1.1.0/) or anything similar?
What is the expected format for a changelog entry?

**Comment by vzarytovskii** [comment]
> Can you please add some guidance on what we are expected to write?
> 
> Are we following [KeepAChangelog](https://keepachangelog.com/en/1.1.0/) or anything similar?
> 
> What is the expected format for a changelog entry?

Not will post a comment with instructions once I sort out auth.

**Comment by nojaf** [comment]
> Bot will post a comment with instructions once I sort out auth.

Hmm, that sounds a bit counterproductive that I would first need to raise the PR, only for a bot to yell at me.

Can the guidance not written in documentation? If I know I need to do this anyway, a clear how-to page will get the job done right?

Also, will the bot validate the entries? Will an entry without a PR link be pointed out for example?

**Comment by T-Gro** [comment]
> > Bot will post a comment with instructions once I sort out auth.
> 
> Hmm, that sounds a bit counterproductive that I would first need to raise the PR, only for a bot to yell at me.
> 
> Can the guidance not written in documentation? If I know I need to do this anyway, a clear how-to page will get the job done right?
> 
> Also, will the bot validate the entries? Will an entry without a PR link be pointed out for example?

There might a circular dependency - we want to have the PR link in there, but before raising the PR, you will not know what the URL is.

As for the content I would like to see:
Write it for the humans who will read it (this will differ across the different changelog files).
* FCS only changes - for FCS API users, so especially tooling authors
* Language changes - write it in a way to explain the language feature to a normal F# user, ideally including motivation and code samples. This can of course reference an existing RFC, does not have to be rewritten.
* ...

**Comment by vzarytovskii** [comment]
> > Bot will post a comment with instructions once I sort out auth.
> 
> 
> 
> Hmm, that sounds a bit counterproductive that I would first need to raise the PR, only for a bot to yell at me.
> 
> 
> 
> Can the guidance not written in documentation? If I know I need to do this anyway, a clear how-to page will get the job done right?
> 
> 
> 

It will also be in the PR temple as a footer. And optionally in separate document.

> 
> Also, will the bot validate the entries? Will an entry without a PR link be pointed out for example?
> 

That's the plan. But first I need to sort out authorization.

**Comment by smoothdeveloper** [comment]
> There might a circular dependency

Not to distract from release note process (amazing!), but can we sort out the circular dependency with:

* approved suggestion
* creating a RFC with the ID with merged place holder file in the lang-design repo
* creating a discussion with links to suggestion, RFC and place holder or link to PR that points to the RFC
* asking the author of suggestion to put links to RFC, discussion and PR (this should be under a certain consistent layout, at the top of the suggestion top post, otherwise people may not head there
* closing the suggestion thread with a link to the RFC discussion one

This is a bit of a tedious.

Regarding this Release note PR:

> let's make everyone's life miserable and 

When I look at the pipeline, and why / how it is done, I'll not feel miserable filling the release notes for my contributions, this is amazing effort :)

**Comment by vzarytovskii** [comment]
Alright, we have a chicken-egg situation here. YAML file should be in `main` in order to apply correct permissions. But to get it to main we need to make sure it works as we want it to, so what I'm proposing is to see if everything works fine here: https://github.com/vzarytovskii/fsharp/pull/7 and if it is, I will force merge it and we'll make a test PR to test it "properly"


Couple of things missing:
1. PR URL check in message.
2. PR template

cc @nojaf

**Comment by vzarytovskii** [comment]
Have link checker working. Will continue with opt-out things and PR template later.

**Comment by 0101** [comment]
Can we just add it as optional in the beginning for testing purposes?

**Comment by vzarytovskii** [comment]
> Can we just add it as optional in the beginning for testing purposes? 

No, everyone must suffer. Also, there will be an easy option-in (with the name and shame comment).

**Comment by vzarytovskii** [comment]
> If there's a way to automatically add the links, then it sounds good. I know there're requests to make this info available, but does having the links worth it to make opening PRs less convenient for the actual contributors? I think we should consider these implications.
>

We can't automate it, unfortunately. Can't run commits based from workflow runs.
 
> Another approach we could consider is we could add the links with a script once per release cycle, based on `git blame`.

We don't own majority of the release process, as well as blames can be inaccurate because of merge conflicts.


I don't have strong opinion about having links, but it was requested by multiple people and in general quite handy.

Usually PRs which require release notes are not done with just one commit, so updating it might be as much of a chore as people think.

It takes a lot (and I mean **A LOT**) of time for us to prepare posts/release notes even for small releases (and that's the main reason nobody wanted to do it). Having release notes will help a lot.

I can make having PR link a warning in the comment for bot, and we'll see from there. 

Optionally, we can require one of the links - either issue or PR (so, if PR fixes an issue, it will be enough to specify the issue and should be easy to track back to PR).

**Comment by vzarytovskii** [comment]
@dotnet/fsharp-team-msft This needs to be merged in order for it to work (I can disable it in UI in case of any failures).

**Comment by vzarytovskii** [comment]
I also made an opt-out by just using tag, so even if something is failing, adding tag _should_ only re-run this singular check.

File: azure-pipelines.yml [hunk 1]
```diff
@@ -283,7 +283,7 @@ stages:
         # We will try to restore both FSharp.Core and FCS and if restore is _successful_, package version needs to be bumped.
         # NOTE: This CI check should only run on the release branches.
         - job: Check_Published_Package_Versions
-          condition: or(startsWith(variables['Build.SourceBranch'], 'refs/heads/release/'), or(startsWith(variables['System.PullRequest.SourceBranch'], 'release/dev'), startsWith(variables['System.PullRequest.TargetBranch'], 'release/dev')))
+          # condition: or(startsWith(variables['Build.SourceBranch'], 'refs/heads/release/'), or(startsWith(variables['System.PullRequest.SourceBranch'], 'release/dev'), startsWith(variables['System.PullRequest.TargetBranch'], 'release/dev')))
```
Comment (0101) [inline azure-pipelines.yml:286]: Should this go back?

**Review by 0101 — APPROVED** [review]

File: azure-pipelines.yml (hunk 1 above)
Comment (vzarytovskii) [reply azure-pipelines.yml:286]: Yep, will add the comment, thanks.

**Comment by vzarytovskii** [comment]
I intend to merge it to main for it to be part of CI. There's an easy way to opt-out as well as relaxed rules for the format for now.

I think it should "go live" in order to understand what's working and what's not.

**Comment by vzarytovskii** [comment]
Alright, since there are no objections...

**Pull request merged**